### DISCUSSION TITLE
Playing multiple dialogues at the same time

### DISCUSSION THREAD

**Discussion by Strate-S** [dialogic-godot/dialogic#2113] [Q&A] [answered]
I have been experimenting for a few weeks but I've been stuck on this matter: is there a way to play multiple dialogues at the same time? Imagine a speech bubble style. Now consider that there are two characters on screen. What if those two characters spoke at the same time, each with their own speech bubble? Two speech bubbles in total, both played at the same time, but each one with different text and from a different character. Is this possible to do? Attached an image illustrating what I mean.
![multiple dialogues](https://github.com/dialogic-godot/dialogic/assets/75809068/4405a788-b66e-429b-bc90-d5d795cb0945)

**Comment by Invertex** [comment]
Out of the box I don't see a simple way you could do this.
If the conversation between the two of them is a written scene, and you just want the ability to have some overlapping call/response, then you could maybe create some sort of custom timeline event, named something like "Multi Text", which would have an array of strings like:

- Jesse: Blah blah!
- James: Bluh buhhh!!
- Cat: Meow

And in your custom dialogue layout, you could have a receiving function like `multi_text_display(string_array)`. And this would parse the lines and feed each of the lines to the appropriate character's bubble typewriters, activating it.

Would probably have to modify some of the scripting for your layout as well to properly handle some signals/events.

**Reply by Strate-S** [reply]
> And this would parse the lines and feed each of the lines to the appropriate character's bubble typewriters, activating it.

So, avoid using the normal timeline's text reproduction and do it through code instead? How would I go about doing this, and would this still allow me to use the usual special functions like changing portraits mid-text and such?

**Reply by Invertex** [reply, marked as answer]
The main thing to look at is creating your own version of `Layer_TextBubble` to use, specifically changing the `text_bubble_base.gd` script so that the `_on_dialogic_text_event(info)` function in it can handle multiple bubbles being open instead of closing all but the singular character being passed in.

A `Dictionary` is passed into this event which can contain all kinds of information. The default `event_text.gd` (which is the "Text" node in the Timeline) only adds a singular `Character` entry into that dictionary (the one you set at the top of the Text node), along with the text info and portrait info.

So there's some different paths you could take here. The most robust would be to replicate the `event_text.gd` class into something like `event_text_multi.gd` that would have multiple character and portrait fields on the header and multiple text boxes for those to write text in. 

And then in the `execute()` function, where the `event_text.gd` function normally passes it's info:
`dialogic.Text.about_to_show_text.emit({'text':final_text, 'character':character, 'portrait':portrait, 'append': is_append})`

You can change that structure to include all your extra characters, portraits and text, which your bubble class's `_on_dialogic_text_event(info)` will then contain in that `info` Dictionary, so you can them control which bubbles do what and how their portraits change.

There will probably be more to it than that, like checking into how Timeline progression happens if two bubbles are open and reacting to Input events... So that might require some modification in that area too.

But this is the general path to getting that kind of system going. I can't really write out much more without basically making the whole system myself haha

**Reply by Strate-S** [reply]
I see, thank you for the detailed response.
